controlplane/kubeadm/internal/workload_cluster.go

```diff
@@ -358,12 +458,53 @@ func checkStaticPodReadyCondition(pod corev1.Pod) error {
 			return errors.Errorf("static pod %s/%s is not ready", pod.Namespace, pod.Name)
 		}
 	}
+
+	// TODO: Remove this as static pods have ready condition.
 	if !found {
 		return errors.Errorf("pod does not have ready condition: %v", pod.Name)
 	}
 	return nil
 }
 
+func checkStaticPodFailedPhase(pod corev1.Pod) bool {
+	return pod.Status.Phase == corev1.PodFailed
+}
+
+// If Pod is in Pending phase and PodScheduled or Initialized condition is set to false, then pod is in provisioning state.
+func checkStaticPodProvisioning(pod corev1.Pod) bool {
+	// Pod being not in Pending phase means it has already provisioned and running or failed or in unknown phase.
+	if pod.Status.Phase != corev1.PodPending {
+		return false
+	}
+
+	for _, condition := range pod.Status.Conditions {
+		if condition.Type == corev1.PodScheduled && condition.Status != corev1.ConditionTrue {
+			return true
+		}
+		if condition.Type == corev1.PodInitialized && condition.Status != corev1.ConditionTrue {
+			return true
+		}
+	}
+	return false
+}
+
+// If Pod is in Pending phase, there may be some failing containers in the pod.
+// This function must be called when PodScheduled and Initialized condition is set to true. If a Pod is initialized but still in Pending state,
+// returns non-nil string with the reason of the container in waiting state.
+func checkStaticPodAfterProvisioningState(pod corev1.Pod) string {
```

Comment:
Should we return an error here instead of a string?

Reply:
This is to surface container-level errors (such as ImagePullBackOff) in the error message, I might have deleted the message during reviews, re-added it.

Reply:
I'd still prefer to get an error instead of a string, for consistency with the other functions